src/test/java/com/fasterxml/jackson/databind/deser/jdk/AtomicReferenceWithStdTypeResolverBuilder4838Test.java

```diff
@@ -0,0 +1,109 @@
+package com.fasterxml.jackson.databind.deser.jdk;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.junit.jupiter.api.Test;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.jsontype.impl.StdTypeResolverBuilder;
+import com.fasterxml.jackson.databind.testutil.DatabindTestUtil;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+// Reported via [modules-java8#86] Cannot read `Optional`s written with `StdTypeResolverBuilder`
+public class AtomicReferenceWithStdTypeResolverBuilder4838Test
+        extends DatabindTestUtil
+{
+
+    public static class Foo<T> {
+        public AtomicReference<T> value;
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Foo<?> foo = (Foo<?>) o;
+            return Objects.equals(value.get(), foo.value.get());
+        }
+    }
+
+    public static class Pojo86 {
+        public String name;
+
+        public static Pojo86 valueOf(String name) {
+            Pojo86 pojo = new Pojo86();
+            pojo.name = name;
+            return pojo;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Pojo86 pojo86 = (Pojo86) o;
+            return Objects.equals(name, pojo86.name);
+        }
+    }
+
+    @JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.WRAPPER_OBJECT)
```

Comment:
If changing `As.WRAPPER_OBJECT` to any other choice -- specifically, `As.PROPERTY` or `As.WRAPPER_ARRAY`, test fails, because of earlier write of TypeId assuming output value is a scalar (need not be JSON String).

Reply:
added [some comment above](https://github.com/FasterXML/jackson-databind/pull/4838#discussion_r1893898841)

Reply:
Quick note: Default Typing does not override `@JsonTypeInfo` since annotation is more specific than general setting. This is why we probably need separate types.
Although it also gets bit more complicated to reason about combinations.

Reply:
True, general one should not take effect if there is more specific one.

Reply:
Reverting test modification